AFQ/viz_libs/afq_plotly.py

```diff
@@ -0,0 +1,552 @@
+import tempfile
+import os
+import os.path as op
+import enum
+import logging
+
+import numpy as np
+
+import dipy.tracking.streamlinespeed as dps
+
+import AFQ.viz_libs.utils as vut
+
+try:
+    import plotly
+    import plotly.graph_objs as go
+except ImportError:
+    raise ImportError(vut.viz_import_msg_error("plotly"))
+
+viz_logger = logging.getLogger("AFQ.viz")
+
+
+def _inline_interact(figure, show, show_inline):
+    """
+    Helper function to reuse across viz functions
+    """
+    if show:
+        viz_logger.info("Creating interactive figure in HTML file...")
+        plotly.offline.plot(figure)
+
+    if show_inline:
+        viz_logger.info("Creating interactive figure inline...")
+        plotly.offline.init_notebook_mode()
+        plotly.offline.iplot(figure)
+
+    return figure
+
+
+def _to_color_range(num):
+    if num < 0:
+        num = 0
+    if num >= 0.999:
+        num = 0.999
+    return num
+
+
+def _color_arr2str(color_arr, opacity=1.0):
+    return (
+        f"rgba({_to_color_range(color_arr[0])}, "
+        f"{_to_color_range(color_arr[1])}, "
+        f"{_to_color_range(color_arr[2])}, "
+        f"{_to_color_range(opacity)})"
+    )
+
+
+def set_layout(figure, color=None):
+    if color is None:
+        color = f"rgba(0,0,0,0)"
+
+    figure.update_layout(
+        plot_bgcolor=color,
+        scene=dict(
+            xaxis=dict(
+                showbackground=False, showticklabels=False, title=''),
+            yaxis=dict(
+                showbackground=False, showticklabels=False, title=''),
+            zaxis=dict(
+                showbackground=False, showticklabels=False, title='')
+        )
+    )
+
+
+def _draw_streamlines(figure, sls, color, name, n_points=100, cbv=None):
+    x_pts = []
+    y_pts = []
+    z_pts = []
+
+    if cbv is not None:
+        customdata = []
+        line_color = []
+        cbv_max = cbv.max()
+        color_max = color.max()
+    else:
+        customdata = None
+        line_color = _color_arr2str(color)
+
+    for sl in sls:
```

Comment:
I think that this could be sped up by using the underlying data in a `StatefulTractogram`. 

Looking at one of these objects, if you do something like this:

```
sft = afd.load_tractogram('fname.trk', 'ref.nii.gz')
```
then, `sft.streamlines._data` is an `(N, 3)` array (with N the number of nodes is all the streamlines), and `sft.streamlines._offsets` is an `(M, )` array (M is the number of streamlinse) where each item is the index of the start of the next streamline (the first item is always `0`, for the start of the first streamline). There is also an attribute `sft.streamlines._lengths`, which provides complementary information. 

So -- I *think* that you could avoid this loop and the rather expensive calls to `extend` below, potentially using only preallocations and indexing operations.

Reply:
Making a StatefulTractogram involves getting a reference / space or limiting possible inputs to only trk filenames. Is that something we want to require here?

Reply:
In the context it's currently used (as a helper function in `visualize_bundles`) these elements are already required, so maybe that's not too burdensome. 

OTOH, we might not even need to require that. `sft.streamlines` is an `ArraySequence` object. So, what if we require the `sls` input here be an `ArraySequence` object? Then it has these attributes (`_data` ,'_offsets' and `lengths`). 

I guess I don't know what `sls` is now. What kind of object can be used as input here?

Reply:
It comes in as:
trk : str, list, or Streamlines
        The streamline information
that gets passed to tract_loader which returns a tractogram, which then gets passed to tract_generator which returns a list of streamlines (the documentation should be more clear here, I will update it), which is used by either fury or plotly to plot.

Considering we only call this from the API, it may make sense to require a stateful tractogram or path to trk file or image instead of just streamlines. However, currently we only require streamlines, no image information required. The question is, would end users directly visualize_bundles find the image requirement annoying/unnecessary? 

If we wanted maximum flexibility, we could change which algorithm we use depending on the information given. IE, if a stateful tractogram is passed in, use its offsets, but if a normal tractogram is passed in, iterate through it.

Reply:
I think that it would be OK to require `StatefulTractogram` as input. We use SFT objects everywhere else.